// @skyway-sdk/common/dist/util.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackOff = void 0;
class BackOff {
    /** @example total 20.4 sec {var sum=0;for(i=0;i<=8;i++){sum+=i**2*100}} */
    constructor(options = {
        duration: 100,
        times: 8,
    }) {
        this.options = options;
        this.retry = 0;
    }
    wait() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.retry++ > this.options.times) {
                return false;
            }
            const timeout = Math.pow(this.retry, 2) * this.options.duration;
            yield new Promise((r) => setTimeout(r, timeout));
            return true;
        });
    }
}
exports.BackOff = BackOff;
