// @skyway-sdk/core/dist/member/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemberImpl = void 0;
const common_1 = require("@skyway-sdk/common");
/**@internal */
class MemberImpl {
    constructor(args) {
        this._status = 'joined';
        this._events = new common_1.Events();
        this.onLeft = this._events.make();
        this.onMetadataUpdated = this._events.make();
        this.channel = args.channel;
        this.id = args.id;
        this.name = args.name;
        this._metadata = args.metadata;
    }
    get metadata() {
        return this._metadata;
    }
    get status() {
        return this._status;
    }
    get publications() {
        return this.channel.publications.filter((p) => p.publisher.id === this.id);
    }
    get subscriptions() {
        return this.channel.subscriptions.filter((p) => p.subscriber.id === this.id);
    }
    toJSON() {
        return {
            id: this.id,
            name: this.name,
            type: this.type,
            subtype: this.subtype,
            metadata: this.metadata,
        };
    }
    /** @private*/
    _left() {
        this._status = 'left';
        this.onLeft.emit();
        this._events.dispose();
    }
    /** @private*/
    _metadataUpdated(metadata) {
        this._metadata = metadata;
        this.onMetadataUpdated.emit(metadata);
    }
    updateMetadata(metadata) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this.channel._updateMemberMetadata(this.id, metadata);
        });
    }
}
exports.MemberImpl = MemberImpl;
