// @skyway-sdk/core/dist/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkyWayConfig = void 0;
const deepmerge_1 = __importDefault(require("deepmerge"));
class SkyWayConfig {
    constructor(options = {}) {
        this.rtcApi = {
            domain: 'b-rapi.beta.skyway.ntt.com',
            timeout: 30000,
            secure: true,
        };
        this.iceParamServer = {
            domain: 'b-ice.beta.skyway.ntt.com',
            version: 1,
            secure: true,
        };
        this.messageService = {
            domain: 'b-msas.beta.skyway.ntt.com',
            secure: true,
        };
        this.rtcConfig = {
            timeout: 30000,
            turnPolicy: 'enable',
            encodedInsertableStreams: false,
        };
        this.token = {
            updateReminderSec: 60,
        };
        this.logLevel = 'error';
        this.debug = { disableDPlane: false };
        Object.assign(this, (0, deepmerge_1.default)(this, options));
    }
    static initialize(options = {}) {
        SkyWayConfig._instance = new SkyWayConfig(options);
        return SkyWayConfig._instance;
    }
    static get get() {
        if (!SkyWayConfig._instance)
            throw new Error('Config not initialized');
        return SkyWayConfig._instance;
    }
}
exports.SkyWayConfig = SkyWayConfig;
