// @skyway-sdk/room/dist/member/local/base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalRoomMemberImpl = void 0;
const common_1 = require("@skyway-sdk/common");
const __1 = require("..");
/**@internal */
class LocalRoomMemberImpl extends __1.RoomMemberImpl {
    /**@private */
    constructor(member, room) {
        super(member, room);
        this.side = 'local';
        this._local = this._member;
        this.onStreamPublished = new common_1.Event();
        this.onStreamUnpublished = new common_1.Event();
        this.onPublicationChanged = new common_1.Event();
        this.onStreamSubscribed = new common_1.Event();
        this.onStreamUnsubscribed = new common_1.Event();
        this.onSubscriptionChanged = new common_1.Event();
        this._local.onStreamSubscribed.add((e) => __awaiter(this, void 0, void 0, function* () {
            const roomSubscription = room._addSubscription(e.subscription);
            this.onStreamSubscribed.emit({
                subscription: roomSubscription,
                stream: e.stream,
            });
        }));
        this._listenRoomEvent();
        this.onStreamPublished.add(() => this.onPublicationChanged.emit());
        this.onStreamUnpublished.add(() => this.onPublicationChanged.emit());
        this.onStreamSubscribed.add(() => this.onSubscriptionChanged.emit());
        this.onStreamUnsubscribed.add(() => this.onSubscriptionChanged.emit());
    }
    _listenRoomEvent() {
        this.room.onStreamUnsubscribed.add((e) => {
            if (e.subscription.subscriber._member.id ===
                this._local.id) {
                this.onStreamUnsubscribed.emit(e);
            }
        });
    }
    leave() {
        return this._local.leave();
    }
}
exports.LocalRoomMemberImpl = LocalRoomMemberImpl;
