// @skyway-sdk/room/dist/member/local/p2p.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalP2PRoomMemberImpl = void 0;
const common_1 = require("@skyway-sdk/common");
const core_1 = require("@skyway-sdk/core");
const base_1 = require("./base");
const log = new common_1.Logger('packages/room/src/member/local/p2p.ts');
/**@internal */
class LocalP2PRoomMemberImpl extends base_1.LocalRoomMemberImpl {
    /**@private */
    constructor(member, room) {
        super(member, room);
    }
    publish(stream, options = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            const publication = yield this._local.publish(stream, options);
            const roomPublication = this.room._addPublication(publication);
            this.onStreamPublished.emit({ publication: roomPublication });
            return roomPublication;
        });
    }
    unpublish(publicationId) {
        return __awaiter(this, void 0, void 0, function* () {
            this._local.unpublish(publicationId);
            const { publication } = yield this.room.onStreamUnpublished
                .watch((e) => e.publication.id === publicationId, core_1.SkyWayConfig.get.rtcApi.timeout)
                .catch(() => {
                throw new common_1.SkyWayError({
                    type: 'timeout',
                    message: 'onStreamUnpublished',
                });
            });
            this.onStreamUnpublished.emit({ publication });
        });
    }
    subscribe(publicationId) {
        return __awaiter(this, void 0, void 0, function* () {
            const { subscription, stream } = yield this._local.subscribe(publicationId);
            const roomSubscription = this.room._addSubscription(subscription);
            return {
                subscription: roomSubscription,
                stream: stream,
            };
        });
    }
    unsubscribe(subscriptionId) {
        return __awaiter(this, void 0, void 0, function* () {
            this._local.unsubscribe(subscriptionId);
            yield this.room.onStreamUnsubscribed
                .watch((e) => e.subscription.id === subscriptionId, core_1.SkyWayConfig.get.rtcApi.timeout)
                .catch(() => {
                throw new common_1.SkyWayError({
                    type: 'timeout',
                    message: 'onStreamUnsubscribed',
                });
            });
        });
    }
    _updateRoom(room) {
        log.debug('_updateRoom', { memberId: this.id });
        this.room = room;
        this._listenRoomEvent();
    }
}
exports.LocalP2PRoomMemberImpl = LocalP2PRoomMemberImpl;
