// @skyway-sdk/token/dist/encoder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SkyWayAuthToken = void 0;
const common_1 = require("@skyway-sdk/common");
const jwt_decode_1 = __importDefault(require("jwt-decode"));
const jwt_encode_1 = __importDefault(require("jwt-encode"));
class SkyWayAuthToken {
    constructor(props) {
        Object.assign(this, props);
    }
    static Decode(token) {
        try {
            const props = (0, jwt_decode_1.default)(token);
            const authToken = new SkyWayAuthToken(props);
            authToken.tokenString = token;
            return authToken;
        }
        catch (error) {
            throw new common_1.SkyWayError({
                type: 'invalidParameter',
                message: 'failed to decode token',
                payload: error,
            });
        }
    }
    encode(secret) {
        this.tokenString = (0, jwt_encode_1.default)({ jti: this.jti, iat: this.iat, exp: this.exp, scope: this.scope }, secret);
        return this.tokenString;
    }
}
exports.SkyWayAuthToken = SkyWayAuthToken;
