// @skyway-sdk/core/dist/media/stream/local/audio.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalAudioStream = void 0;
const common_1 = require("@skyway-sdk/common");
const util_1 = require("../../util");
const share_1 = require("../share");
const _1 = require(".");
class LocalAudioStream extends _1.LocalStreamBase {
    constructor(label, track) {
        super(label, 'audio');
        this.track = track;
        this.contentType = 'audio';
        /**@private */
        this._mutedTrack = (0, util_1.silence)();
        /**@private */
        this._onReplaceTrack = new common_1.Event();
    }
    /**
     * @description {english} Attach the stream to the element.
     */
    attach(element) {
        this._element = element;
        (0, share_1.attachElement)(element, this.track);
    }
    /**
     * @description {english} Detach the stream from the element.
     */
    detach() {
        if (this._element) {
            (0, share_1.detachElement)(this._element, this.track);
            this._element = undefined;
        }
    }
    replaceTrack(track) {
        this._onReplaceTrack.emit(track);
    }
}
exports.LocalAudioStream = LocalAudioStream;
