// @skyway-sdk/core/dist/media/stream/local/video.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalVideoStream = void 0;
const common_1 = require("@skyway-sdk/common");
const share_1 = require("../share");
const _1 = require(".");
const util_1 = require("./util");
class LocalVideoStream extends _1.LocalStreamBase {
    constructor(label, track) {
        super(label, 'video');
        this.track = track;
        this.contentType = 'video';
        /**@private */
        this._mutedTrack = (0, util_1.blankScreen)({
            width: this.track.getSettings().width,
            height: this.track.getSettings().height,
        });
        /**@private */
        this._onReplaceTrack = new common_1.Event();
    }
    /**
     * @description {english} Attach the stream to the element.
     */
    attach(element) {
        this._element = element;
        (0, share_1.attachElement)(element, this.track);
    }
    /**
     * @description {english} Detach the stream from the element.
     */
    detach() {
        if (this._element) {
            (0, share_1.detachElement)(this._element, this.track);
            this._element = undefined;
        }
    }
    replaceTrack(track) {
        this._onReplaceTrack.emit(track);
    }
}
exports.LocalVideoStream = LocalVideoStream;
