// @skyway-sdk/rtc-api-client/dist/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RtcApiClient = void 0;
const common_1 = require("@skyway-sdk/common");
const common_2 = require("@skyway-sdk/common");
const rtc_rpc_api_client_1 = require("@skyway-sdk/rtc-rpc-api-client");
const _1 = require(".");
const config_1 = require("./config");
const api_1 = require("./infrastructure/api");
const log = new common_1.Logger('packages/rtc-api-client/src/client.ts');
class RtcApiClient {
    constructor(appId, _rpc) {
        this.appId = appId;
        this._rpc = _rpc;
        this.closed = false;
        this.onFatalError = new common_2.Event();
        this.apiClient = new api_1.SkyWayApiImpl(_rpc);
        _rpc.onFatalError.once((e) => {
            this.onFatalError.emit(e);
        });
    }
    static Create(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const config = config_1.Config.initialize(args);
            common_1.Logger.level = config.logLevel;
            log.debug('RtcApiClient spawned', config);
            const api = new rtc_rpc_api_client_1.RtcRpcApiClient(Object.assign(Object.assign({}, config.rtcApi), { token: args.token, logLevel: config.logLevel }));
            yield api.connect();
            return new RtcApiClient(args.appId, api);
        });
    }
    get received() {
        return this._rpc._rpc.received;
    }
    updateAuthToken(token) {
        this._rpc.updateToken(token);
    }
    createChannel(init = {}) {
        return __awaiter(this, void 0, void 0, function* () {
            log.debug('[start] apiClient.createChannel', { init });
            const channelDto = yield this.apiClient
                .createChannel(this.appId, init)
                .catch((e) => {
                log.debug('[failed] apiClient.createChannel', { init, e });
                throw e;
            });
            // todo fix ?
            channelDto.name = init.name;
            log.debug('[end] apiClient.createChannel', { init });
            const channel = (0, _1.channelFactory)(this.appId, this._rpc, this.apiClient, channelDto);
            return channel;
        });
    }
    findChannel(query) {
        return __awaiter(this, void 0, void 0, function* () {
            log.debug('[start] apiClient.getChannel', { query });
            const channelDto = yield this.apiClient
                .getChannel(this.appId, query)
                .catch((e) => {
                log.debug('[failed] apiClient.getChannel', { query, e });
                throw e;
            });
            const channel = (0, _1.channelFactory)(this.appId, this._rpc, this.apiClient, channelDto);
            log.debug('[end] apiClient.getChannel', { channelId: channel.id });
            return channel;
        });
    }
    findOrCreateChannel(query) {
        return __awaiter(this, void 0, void 0, function* () {
            log.debug('[start] apiClient.findOrCreateChannel', { query });
            const channelDto = yield this.apiClient
                .findOrCreateChannel(this.appId, query)
                .catch((e) => {
                log.debug('[failed] apiClient.findOrCreateChannel', { query, e });
                throw e;
            });
            log.debug('[end] apiClient.findOrCreateChannel', { query });
            const channel = (0, _1.channelFactory)(this.appId, this._rpc, this.apiClient, channelDto);
            return channel;
        });
    }
    deleteChannel(channelId) {
        return this.apiClient.deleteChannel(this.appId, channelId);
    }
    close() {
        if (this.closed) {
            return;
        }
        this.closed = true;
        log.debug('closed', { appid: this.appId });
        this.apiClient.close();
    }
}
exports.RtcApiClient = RtcApiClient;
